# A0: jQMM tests notebook

# Introduction

This notebook contains all unit tests fof the jQMM library compiled into a single notebook, for ease of use.

# Setup

First set environment and path variables properly:

In [1]:
quantmodelDir = '/users/hgmartin/libraries/quantmodel'

This is the only place where the jQMM library path needs to be set.

In [2]:
%matplotlib inline

import sys, os
pythonPath = quantmodelDir+"/code/core"
if pythonPath not in sys.path:
    sys.path.append(pythonPath)
os.environ["QUANTMODELPATH"] = quantmodelDir

Then import needed modules:

In [3]:
import core, ReactionNetworks, FluxModels, GAMSclasses, ToyaData
import unittest

In [4]:
cd /scratch/hgmartin_scratch/tests

/scratch/hgmartin_scratch/tests


# Tests

## Quick tests

These tests are quite quick and can be run every time a change is made. However, they do not check the full functionality that only the full tests below cover.

### GAMSclasses tests

In [5]:
GAMStest = unittest.TestSuite()

GAMStest.addTest(GAMSclasses.testGAMSInstallation('testGAMS'))

GAMStest.addTest(GAMSclasses.testGAMSProblem('testProblem'))

GAMStest.addTest(GAMSclasses.testGAMSBatch('testBatch'))



unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(GAMStest)

testGAMS (GAMSclasses.testGAMSInstallation)
Checks if gams can be called at the command line successfully ... ok
testProblem (GAMSclasses.testGAMSProblem)
Tests the classic transport problem in GAMS ... ok
testBatch (GAMSclasses.testGAMSBatch)
Tests the classic transport problem in GAMS in batch mode ... ok

----------------------------------------------------------------------
Ran 3 tests in 1.379s

OK


<unittest.runner.TextTestResult run=3 errors=0 failures=0>

### core tests

In [6]:
COREtest = unittest.TestSuite()

COREtest.addTest(core.testMetabolites('testMetabolite'))
COREtest.addTest(core.testMetabolites('testReactantProduct'))

COREtest.addTest(core.testReactions('testPrint'))
COREtest.addTest(core.testReactions('testDicts'))

COREtest.addTest(core.testEmus('testEMU'))
COREtest.addTest(core.testEmus('testEMUtransition'))

COREtest.addTest(core.testAtomTransitions('testfindEMUtransition'))
COREtest.addTest(core.testAtomTransitions('testfindEMUs'))
COREtest.addTest(core.testAtomTransitions('testgetOriginDictionary'))

COREtest.addTest(core.testRangedNumbers('testAll'))

COREtest.addTest(core.testFlux('testAll'))



unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(COREtest)

testMetabolite (core.testMetabolites)
Testing metabolite class ... ok
testReactantProduct (core.testMetabolites)
Testing reactant, product classes ... ok
testPrint (core.testReactions)
Tests reaction printing ... ok
testDicts (core.testReactions)
Tests reaction printing ... ok
testEMU (core.testEmus)
tests EMU class ... ok
testEMUtransition (core.testEmus)
tests EMUTransition class ... ok
testfindEMUtransition (core.testAtomTransitions) ... ok
testfindEMUs (core.testAtomTransitions) ... ok
testgetOriginDictionary (core.testAtomTransitions) ... ok
testAll (core.testRangedNumbers) ... ok
testAll (core.testFlux) ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.009s

OK


<unittest.runner.TextTestResult run=11 errors=0 failures=0>

### ReactionNetworks tests

In [7]:
RNtest = unittest.TestSuite()

RNtest.addTest(ReactionNetworks.testRN('testRxnsMets'))
RNtest.addTest(ReactionNetworks.testRN('testAddReaction'))
RNtest.addTest(ReactionNetworks.testRN('testChangeFluxBounds'))

RNtest.addTest(ReactionNetworks.testC13RN('testAddFeedLabel'))

RNtest.addTest(ReactionNetworks.testTSRN('testAddtransitions'))

RNtest.addTest(ToyaData.testInputFiles('testFull'))
#RNtest.addTest(ToyaData.testInputFiles('refreshSavedFiles'))

RNtest.addTest(ToyaData.testOptimizationFiles('testFull'))
#RNtest.addTest(ToyaData.testOptimizationFiles('refreshSavedFiles'))



unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(RNtest)

testRxnsMets (ReactionNetworks.testRN)
Testing expected reactions and metabolites in basic usage ... ok
testAddReaction (ReactionNetworks.testRN)
Testing result of adding reactions ... ok
testChangeFluxBounds (ReactionNetworks.testRN)
Testing changeFluxBounds in basic usage ... ok
testAddFeedLabel (ReactionNetworks.testC13RN)
Testing Adding feed and labeling ... ok
testAddtransitions (ReactionNetworks.testTSRN)
Testing adding transitions to genome-scale ... ok
testFull (ToyaData.testInputFiles)
Tests all of the strains ... ok
testFull (ToyaData.testOptimizationFiles)
Tests all of the strains ... ok

----------------------------------------------------------------------
Ran 7 tests in 189.815s

OK


<unittest.runner.TextTestResult run=7 errors=0 failures=0>

### FluxModels tests


In [8]:
FMtest = unittest.TestSuite()
 
FMtest.addTest(FluxModels.FBAtests('testFBA'))
FMtest.addTest(FluxModels.FBAtests('testFVA'))

FMtest.addTest(FluxModels.C13MFAtests('testTCAtoy'))
FMtest.addTest(FluxModels.C13MFAtests('testToya2010'))

FMtest.addTest(ToyaData.testTSFluxCalculation('testQuick'))
#FMtest.addTest(ToyaData.testTSFluxCalculation('refreshSavedFiles'))
FMtest.addTest(ToyaData.testELVA('testQuick'))
FMtest.addTest(ToyaData.testConstPower('testQuick'))


unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(FMtest)

testFBA (FluxModels.FBAtests)
Checks results from FBA ... ok
testFVA (FluxModels.FBAtests)
Checks results from FVA ... ok
testTCAtoy (FluxModels.C13MFAtests)
Checks results from TCA toy model ... ok
testToya2010 (FluxModels.C13MFAtests)
Checks results using Toya 2010 data ... ok
testQuick (ToyaData.testTSFluxCalculation)
Only tests the first of the strains ... 
Testing strain: wt5h

ok
testQuick (ToyaData.testELVA)
Quick test for wt5h only ... ok
testQuick (ToyaData.testConstPower)
Quick test for wt5h only ... ok

----------------------------------------------------------------------
Ran 7 tests in 4361.318s

OK


<unittest.runner.TextTestResult run=7 errors=0 failures=0>

### Predictions

In [9]:
Ptest = unittest.TestSuite()
 
Ptest.addTest(ToyaData.testPredictions('testQuick'))
#Ptest.addTest(ToyaData.testPredictions('refreshSavedFiles'))

unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(Ptest)

testQuick (ToyaData.testPredictions)
Quick test for pyk5h only ... ok

----------------------------------------------------------------------
Ran 1 test in 1292.293s

OK


<unittest.runner.TextTestResult run=1 errors=0 failures=0>

## Full tests

### FluxModels tests

In [10]:
FMtest = unittest.TestSuite()
 
FMtest.addTest(ToyaData.testTSFluxCalculation('testFull'))

unittest.TextTestRunner(stream=sys.stdout,verbosity=2).run(FMtest)

testFull (ToyaData.testTSFluxCalculation)
Test all strains ... 
Testing strain: wt5h


Testing strain: wt6h


Testing strain: wt7h


Testing strain: pyk5h


Testing strain: pyk6h


Testing strain: pyk7h


Testing strain: pgi16h


Testing strain: pgi21h


Testing strain: pgi23h

FAIL

FAIL: testFull (ToyaData.testTSFluxCalculation)
Test all strains
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/users/hgmartin/libraries/quantmodel/code/core/ToyaData.py", line 294, in testFull
    self.assertTrue(AllCasesWork,msg=errmsg)
AssertionError: 
Strain:pyk5h

Flux AKGDH changed from: 
[0:0.36471019:0.99286] to: 
[0:0:1.11912]
Flux EX_fum_e_ changed from: 
[0:0:2.1368] to: 
[0:0:2.53215]
Flux THD2 changed from: 
[1.95049:2.69433729906:17.71472] to: 
[1.75855:2.98213982546:19.07007]



----------------------------------------------------------------------
Ran 1 test in 9910.972s

FAILED (failures=1)


<unittest.runner.TextTestResult run=1 errors=0 failures=1>

Because of the stochastic nature of the optimization algorithm, some results are not always repeated exactly. In this case AKGDH flux changed from [0, 0.36, 0.99] \[lowest compatible with 13C data, best fit, highest\] to [0, 0, 1.11]. This is not terribly unexpected: the flux value is expected to be anything between 0.0 and 0.99, so a shift from 0.36 to 0.0 is not surprising. Also, the upper bound changed from 0.99 to 1.11 which is a bit more than the 10% allowed by the unit test before calling it a failed test. That 10% is arbitrary, a 15% would be equally acceptable given the stochastic nature of the algorithm. Similar reasoning can be applied to EX_fum_e and THD2.